### PR TITLE
Support reading local/hdfs/s3 file compressed with snappy

### PR DESCRIPTION
**Pull request by taiyang-li**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support reading local/hdfs/s3 file compressed with snappy

Detailed description / Documentation draft:
HDFS files in hive table maybe compressed with snappy algorithm.  By implementing `SnappyReadBuffer`, Clickhouse is able to read data from those files.

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Update ya.make files via `utils/generate_yamake`

**Comment by taiyang-li** [comment]
> Update ya.make files via `utils/generate_yamake`

Already done. But git remind me `Everything is up-to-date`.

File: src/IO/examples/snappy_read_buffer.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include <IO/ReadBufferFromString.h>
+#include <IO/SnappyReadBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <IO/copyData.h>
+#include "Common/Exception.h"
+
+
+int main()
+{
+    using namespace DB;
```
Comment (kssenii) [inline src/IO/examples/snappy_read_buffer.cpp:15]: Can you please also add an integration test where snappy will be passed as compression method?
For example, if you have a ready compressed file, you can put it into container inside an integration test 
like here https://github.com/ClickHouse/ClickHouse/blob/9e496910a0ddbe7d75b911d2a5a9b34c75b852d6/tests/integration/test_library_bridge/test.py#L37 and then put it into hdfs via `pyhdfs` command `copy_from_local` .
example tests with pyhdfs examples are here https://github.com/ClickHouse/ClickHouse/blob/9e496910a0ddbe7d75b911d2a5a9b34c75b852d6/tests/integration/test_merge_tree_hdfs/test.py#L167
But the test should be put it `test_storage_hdfs/test.py`
Or similar can ba done with s3.

Comment (taiyang-li) [reply src/IO/examples/snappy_read_buffer.cpp:15]: ok.

Comment (taiyang-li) [reply src/IO/examples/snappy_read_buffer.cpp:15]: @kssenii functional test already done, pls refer to `tests/queries/0_stateless/01059_storage_file_compression.sh`

File: src/IO/CMakeLists.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
-if (ENABLE_EXAMPLES)
+# if (ENABLE_EXAMPLES)
     add_subdirectory (examples)
-endif ()
+# endif ()
```
Comment (kssenii) [inline src/IO/CMakeLists.txt:3]: Need to uncomment back.

File: src/IO/CompressionMethod.cpp [hunk 3]
```diff
@@ -139,6 +149,12 @@ std::unique_ptr<WriteBuffer> wrapWriteBufferWithCompressionMethod(
     if (method == CompressionMethod::Bzip2)
         return std::make_unique<Bzip2WriteBuffer>(std::move(nested), level, buf_size, existing_memory, alignment);
 #endif
+
+#if USE_SNAPPY
+    if (method == CompressionMethod::Snappy)
+        return std::make_unique<SnappyWriteBuffer>(std::move(nested), buf_size, existing_memory, alignment);
```
Comment (kssenii) [inline src/IO/CompressionMethod.cpp:155]: We can also pass and set compression level parameter as for other compression methods.

File: src/IO/SnappyReadBuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+}
+
+SnappyReadBuffer::SnappyReadBuffer(std::unique_ptr<ReadBuffer> in_, size_t buf_size, char * existing_memory, size_t alignment)
+    : BufferWithOwnMemory<SeekableReadBuffer>(buf_size, existing_memory, alignment), in(std::move(in_))
+{
+    WriteBufferFromString wb(compress_buffer);
+    copyData(*in, wb);
+
+    bool success = snappy::Uncompress(compress_buffer.data(), wb.count(), &uncompress_buffer);
```
Comment (kssenii) [inline src/IO/SnappyReadBuffer.cpp:30]: This is not correct to do it this way and also not correct to do it in constructor. Need to override `nextImpl` and implement this logic like in other compression read buffers.

File: src/IO/examples/snappy_read_buffer.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/IO/examples/snappy_read_buffer.cpp:15]: done.

File: src/IO/CMakeLists.txt (hunk 2 above)
Comment (taiyang-li) [reply src/IO/CMakeLists.txt:3]: Done.

File: src/IO/SnappyReadBuffer.cpp (hunk 4 above)
Comment (taiyang-li) [reply src/IO/SnappyReadBuffer.cpp:30]: ok.

Comment (taiyang-li) [reply src/IO/SnappyReadBuffer.cpp:30]: I found it difficult to implement stream compress/decompress for snappy.  As far as I know, snappy library (https://github.com/google/snappy) doesn't contain stream compress/decompress API for snappy. Do you have any ideas about this problem? @kssenii

**Comment by PIGSYPEACH** [comment]
IO/ya.make Has been removed？！

**Comment by kssenii** [comment]
Yes, it was removed from master, @taiyang-li can you please merge with master and remove it from diff?

**Comment by taiyang-li** [comment]
> Yes, it was removed from master, @taiyang-li can you please merge with master and remove it from diff?

Ok.

**Comment by taiyang-li** [comment]
> > Yes, it was removed from master, @taiyang-li can you please merge with master and remove it from diff?
> 
> Ok.

Done

**Review by kssenii — APPROVED** [review]
Still doubtful about using compress/decompress whole file at once, but regarding 
> My PR is implementing the 1st codec. Using raw snappy codec, the data from nested read buffer inside SnappyReadBuffer can't be uncomressed streamly.
Later I will implement another two snappy codec.

I think for now it is ok.

Can you please fix `ClickHouse build check `, `ClickHouse special build check` and I will merge.

**Comment by taiyang-li** [comment]
Contained in  https://github.com/ClickHouse/ClickHouse/pull/31104. Close now

**Pull request closed without merging**